community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/TransactionIdTracking.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.bolt.v1.runtime.internal;
+
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+import org.neo4j.function.Predicates;
+import org.neo4j.kernel.api.exceptions.TransactionFailureException;
+import org.neo4j.kernel.impl.transaction.log.TransactionIdStore;
+
+import static org.neo4j.kernel.impl.transaction.log.TransactionIdStore.BASE_TX_ID;
+
+class TransactionIdTracking implements SessionStateMachine.SPI.TransactionIdTracker
+{
+    private final Supplier<TransactionIdStore> transactionIdStore;
+    private final int timeout;
+    private final TimeUnit timeoutUnit;
+
+    private long txId;
+
+    TransactionIdTracking( Supplier<TransactionIdStore> transactionIdStore, long transactionId, int timeout,
+                           TimeUnit unit )
+    {
+        this.transactionIdStore = transactionIdStore;
+        this.txId = transactionId;
+        this.timeout = timeout;
+        this.timeoutUnit = unit;
+    }
+
+    @Override
+    public void assertUpToDate() throws TransactionFailureException
+    {
+        if ( txId <= BASE_TX_ID )
+        {
+            return;
+        }
+
+        try
+        {
+            Predicates.await( () -> txId <= transactionIdStore.get().getLastClosedTransactionId(), timeout, timeoutUnit,
```

Comment:
I don't understand how you can claim here we use `Exception`s for control flow, do you think it is better  to duplicated the code and write the same `while` loop here instead of using `Predicates.await`?

I'll remove the caught exception as a cuase since it is not relevant and I'll improve the message of the exception we throw.